### PR TITLE
use char instead of &str for single char patterns

### PR DESCRIPTION
**Pull request by matthiaskrgr**

### CODE REVIEW

File: src/librustc_builtin_macros/format.rs [hunk 1]
```diff
@@ -894,7 +894,7 @@ pub fn expand_preparsed_format_args(
     };
 
     let (is_literal, fmt_snippet) = match ecx.source_map().span_to_snippet(fmt_sp) {
-        Ok(s) => (s.starts_with("\"") || s.starts_with("r#"), Some(s)),
+        Ok(s) => (s.starts_with('\"') || s.starts_with("r#"), Some(s)),
```
Comment (ecstatic-morse) [inline src/librustc_builtin_macros/format.rs]: No need to escape this.
```suggestion
        Ok(s) => (s.starts_with('"') || s.starts_with("r#"), Some(s)),
```

File: src/librustc_typeck/check/demand.rs [hunk 2]
```diff
@@ -404,7 +404,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 {
                     if let hir::ExprKind::Lit(_) = expr.kind {
                         if let Ok(src) = sm.span_to_snippet(sp) {
-                            if src.starts_with("\"") {
+                            if src.starts_with('\"') {
```
Comment (ecstatic-morse) [inline src/librustc_typeck/check/demand.rs]: Same as above.

**Review by ecstatic-morse — CHANGES_REQUESTED** [review]
r=me after removing unnecessary escapes.

**Comment by matthiaskrgr** [comment]
diff: https://github.com/rust-lang/rust/compare/b974a814870359c5957c554be18291b71f7758f2..8ceb9096ac72bab8b5566f22645c16e84634603f

r? @ecstatic-morse

**Review by ecstatic-morse — APPROVED** [review]

**Comment by Dylan-DPC** [comment]
failed in rollup 
@bors r-

**Comment by petrochenkov** [comment]
https://github.com/rust-lang/rust/issues/41993 was fixed, but I still think this replacement is ideologically wrong.
All these places don't care about unicode code points, they do substring search, it's just the substring happens to be pretty short.
`char` is a pretty rare data type that shouldn't be used unless you want unicode code points specifically.

**Comment by matthiaskrgr** [comment]
When running these benchmarks that I found via the discussion you linked
````rust
#![feature(test)]

extern crate test;
use crate::test::black_box;
use crate::test::Bencher;

fn main() {
    println!("Hello, world!");
}

#[bench]
fn starts_with_char(b: &mut Bencher) {
    let text = black_box("kdjsfhlakfhlsghlkvcnljknfqiunvcijqenwodind");
    b.iter(|| {
        for _ in 0..1024 {
            black_box(text.starts_with('k'));
        }
    })
}

#[bench]
fn starts_with_str(b: &mut Bencher) {
    let text = black_box("kdjsfhlakfhlsghlkvcnljknfqiunvcijqenwodind");
    b.iter(|| {
        for _ in 0..1024 {
            black_box(text.starts_with("k"));
        }
    })
}

#[bench]
fn ends_with_char(b: &mut Bencher) {
    let text = black_box("kdjsfhlakfhlsghlkvcnljknfqiunvcijqenwodind");
    b.iter(|| {
        for _ in 0..1024 {
            black_box(text.ends_with('k'));
        }
    })
}

#[bench]
fn ends_with_str(b: &mut Bencher) {
    let text = black_box("kdjsfhlakfhlsghlkvcnljknfqiunvcijqenwodind");
    b.iter(|| {
        for _ in 0..1024 {
            black_box(text.ends_with("k"));
        }
    })
}
````
, char is consistently faster for me with default optimization levels:

````
running 4 tests
test ends_with_char   ... bench:         465 ns/iter (+/- 31)
test ends_with_str    ... bench:         544 ns/iter (+/- 79)
test starts_with_char ... bench:         356 ns/iter (+/- 38)
test starts_with_str  ... bench:         691 ns/iter (+/- 10)
````

````
running 4 tests
test ends_with_char   ... bench:         464 ns/iter (+/- 10)
test ends_with_str    ... bench:         544 ns/iter (+/- 15)
test starts_with_char ... bench:         355 ns/iter (+/- 11)
test starts_with_str  ... bench:         692 ns/iter (+/- 11)
````

````
running 4 tests
test ends_with_char   ... bench:         448 ns/iter (+/- 11)
test ends_with_str    ... bench:         527 ns/iter (+/- 2)
test starts_with_char ... bench:         340 ns/iter (+/- 33)
test starts_with_str  ... bench:         670 ns/iter (+/- 8)
````

**Comment by petrochenkov** [comment]
If this is a perf improvement now, rather than just a stylistic change, then it should be ok.

**Comment by matthiaskrgr** [comment]
rebased

**Pull request merged**